File: csharp/EviFluor/verification.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Nodes;

namespace Hse.EviFluor
{
    /// <summary>
    /// Provides verification checks for various measurement types and conditions.
    /// </summary>
    public class Verification
    {
        /// <summary>
        /// Default minimum expected RFU value.
        /// </summary>
        public static readonly double DefaultExpectedMinRfu = 4.5;
        /// <summary>
        /// Minimum expected RFU value used for verification.
        /// </summary>
        public static double ExpectedMinRfu = DefaultExpectedMinRfu;

        /// <summary>
        /// Default maximum expected RFU value.
        /// </summary>
        public static readonly double DefaultExpectedMaxRfu = 35.0;
        /// <summary>
        /// Maximum expected RFU value used for verification.
        /// </summary>
        public static double ExpectedMaxRfu = DefaultExpectedMaxRfu;

        /// <summary>
        /// Default minimum expected LED power.
        /// </summary>
        public static readonly int DefaultExpectedMinLedPower = 32;
        /// <summary>
        /// Minimum expected LED power used for verification.
        /// </summary>
        public static int ExpectedMinLedPower = DefaultExpectedMinLedPower;

        /// <summary>
        /// Default maximum expected LED power.
        /// </summary>
        public static readonly int DefaultExpectedMaxLedPower = 222;
        /// <summary>
        /// Maximum expected LED power used for verification.
        /// </summary>
        public static int ExpectedMaxLedPower = DefaultExpectedMaxLedPower;

        /// <summary>
        /// Default multiplier used for calculating RFU thresholds when determining cuvette presence.
        /// </summary>
        public static readonly double DefaultRfuThresholdMultiplier = 2.0;
        /// <summary>
        /// Multiplier used for calculating RFU thresholds when determining cuvette presence.
        /// </summary>
        public static double RfuThresholdMultiplier = DefaultRfuThresholdMultiplier;

        /// <summary>
        /// Default maximum allowed signal before considering saturation.
        /// </summary>
        public static readonly double DefaultMaxSignal = 2499.0;
        /// <summary>
        /// Maximum allowed signal before considering saturation.
        /// </summary>
        public static double MaxSignal = DefaultMaxSignal;

        /// <summary>
        /// Default target signal value for standard high checks.
        /// </summary>
        public static readonly double DefaultStdHighTarget = 2000.0;
        /// <summary>
        /// Target signal value for standard high checks.
        /// </summary>
        public static double StdHighTarget = DefaultStdHighTarget;

        /// <summary>
        /// Default allowed deviation from the standard high target.
        /// </summary>
        public static readonly double DefaultStdHighDelta = 300;
        /// <summary>
        /// Allowed deviation from the standard high target.
        /// </summary>
        public static double StdHighDelta = DefaultStdHighDelta;

        /// <summary>
        /// Default threshold for detecting negative concentrations.
        /// </summary>
        public static readonly double DefaultThresholdNegativeConcentration = -0.1;
        /// <summary>
        /// Threshold for detecting negative concentrations.
        /// </summary>
        public static double ThresholdNegativeConcentration = DefaultThresholdNegativeConcentration;

        /// <summary>
        /// Enum representing various types of problems that may occur during verification.
        /// </summary>
        public enum ProblemId
        {
            /// <summary>
            /// Indicates that the sensor signal saturated.
            /// </summary>
            SATURATION = 1,

            /// <summary>
            /// A cuvette was expected.
            /// </summary>
            CUVETTE_MISSING = 2,

            /// <summary>
            /// Auto gain function failed.
            /// </summary>
            AUTO_GAIN_RESULT = 5,

            /// <summary>
            /// The signal level is not in the expected range (1700 - 2300 mV)
            /// </summary>
            WRONG_LEVEL = 6,

            /// <summary>
            /// The concentration is less than ThresholdNegativeConcentration
            /// </summary>
            NEGATIVE_CONCENTRATION = 7
        }

        /// <summary>
        /// Flags that provide context-sensitive hints for verification checks.
        /// </summary>
        [Flags]
        public enum Hints
        {
            /// <summary>
            /// No hint.
            /// </summary>
            NONE = 0,

            /// <summary>
            /// A cuvette is expected.
            /// </summary>
            MUST_HAVE_CUVETTE = 1,

            /// <summary>
            /// Signal for standard high (~2000 +- 200 mV) expected.
            /// </summary>
            STD_HIGH = 2
        }

        /// <summary>
        /// Represents a single verification result entry within the <see cref="Verification"/> process.
        /// </summary>
        /// <remarks>
        /// Each entry consists of a <see cref="Verification.ProblemId"/> indicating the type of problem detected,
        /// and an associated data object (e.g., a measurement or result) that caused the issue.
        /// </remarks>
        public class Entry : IJsonSerializable
        {
            /// <summary>
            /// Gets the problem type that this entry represents.
            /// </summary>
            public ProblemId Problem { get; }

            /// <summary>
            /// Gets the data object associated with the problem (e.g., a measurement or result).
            /// </summary>
            public object Data { get; }

            /// <summary>
            /// Initializes a new instance of the <see cref="Entry"/> class.
            /// </summary>
            /// <param name="problem">The specific problem identifier.</param>
            /// <param name="data">The data related to the problem, such as a measurement.</param>
            public Entry(ProblemId problem, object data)
            {
                Problem = problem;
                Data = data;
            }

            /// <summary>
            /// Returns a string that represents the current verification entry.
            /// </summary>
            /// <returns>A string describing the problem type and the associated data.</returns>
            public override string ToString()
            {
                return $"{Problem}({(int)Problem}) {Data}";
            }

            /// <summary>
            /// Converts this entry into a JSON object containing the problem ID, its name, and the serialized data (if available).
            /// </summary>
            /// <returns>A <see cref="JsonNode"/> representing the verification entry.</returns>
            public JsonNode ToJson()
            {
                var json = new JsonObject
                {
                    ["problem_id"] = (int)Problem,
                    ["description"] = Problem.ToString()
                };

                if (Data is IJsonSerializable serializable)
                {
                    json["data"] = serializable.ToJson();
                }

                return json;
            }
        }

        private readonly List<Entry> entries = new();

        /// <summary>
        /// Provides string representation of all entries.
        /// </summary>
        public override string ToString() => $"entries:{string.Join(", ", entries)}";

        /// <summary>
        /// Indicates whether the verification passed without any issues.
        /// </summary>
        public bool Success() => entries.Count == 0;

        /// <summary>
        /// Indicates whether the verification encountered one or more issues.
        /// </summary>
        public bool Failed() => entries.Count > 0;

        /// <summary>
        /// Checks if a specific problem has been detected.
        /// </summary>
        public bool HasProblem(ProblemId id) =>
            entries.Exists(entry => entry.Problem == id);

        /// <summary>
        /// List of all detected problems.
        /// </summary>
        public List<Entry> Entries => entries;

        /// <summary>
        /// Checks the result of an auto-gain operation.
        /// </summary>
        public bool Check(AutoGainResult autoGainResult, Hints hints = Hints.NONE)
        {
            if (!autoGainResult.Found)
            {
                entries.Add(new Entry(ProblemId.AUTO_GAIN_RESULT, autoGainResult));
                return false;
            }
            return true;
        }

        /// <summary>
        /// Checks the validity of a single measurement based on saturation and expected behavior.
        /// </summary>
        public bool Check(SingleMeasurement sm, Hints hints = Hints.NONE)
        {
            bool result = true;

            if (sm.Channel470.Value >= MaxSignal)
            {
                entries.Add(new Entry(ProblemId.SATURATION, sm));
                result = false;
            }

            if (hints.HasFlag(Hints.MUST_HAVE_CUVETTE))
            {
                if (!HasCuvette(sm))
                {
                    entries.Add(new Entry(ProblemId.CUVETTE_MISSING, sm));
                    result = false;
                }
            }

            if (hints.HasFlag(Hints.STD_HIGH))
            {
                if (sm.Channel470.Value < (StdHighTarget - StdHighDelta) || sm.Channel470.Value > (StdHighTarget + StdHighDelta))
                {
                    entries.Add(new Entry(ProblemId.WRONG_LEVEL, sm));
                    result = false;
                }
            }

            return result;
        }

        /// <summary>
        /// Checks the validity of a first air measurement result.
        /// </summary>
        public bool Check(FirstAirMeasurementResult fam, Hints hints = Hints.NONE)
        {
            bool r1 = Check(fam.MinMeasurement, Hints.MUST_HAVE_CUVETTE);
            bool r2 = Check(fam.MaxMeasurement, Hints.MUST_HAVE_CUVETTE);
            return r1 && r2;
        }

        /// <summary>
        /// Checks the validity of a first sample measurement result.
        /// </summary>
        public bool Check(FirstSampleMeasurementResult fsm, Hints hints = Hints.NONE)
        {
            bool r1 = Check(fsm.AutoGainResult, hints);
            bool r2 = Check(fsm.Measurement, Hints.MUST_HAVE_CUVETTE | Hints.STD_HIGH);
            return r1 && r2;
        }

        /// <summary>
        /// Checks the validity of a full measurement (air and sample).
        /// </summary>
        public bool Check(Measurement m, Hints hints = Hints.NONE)
        {
            bool r1 = Check(m.air, Hints.MUST_HAVE_CUVETTE);
            bool r2 = Check(m.sample, hints | Hints.MUST_HAVE_CUVETTE);
            return r1 && r2;
        }

        /// <summary>
        /// Validates the computed concentration against negative thresholds.
        /// </summary>
        /// <returns><c>true</c> if concentration ≥ <see cref="ThresholdNegativeConcentration"/>; otherwise <c>false</c>.</returns>
        public bool Check(Results results, Hints hints = Hints.NONE)
        {
            if (results.Concentration < ThresholdNegativeConcentration)
            {
                entries.Add(new Entry(ProblemId.NEGATIVE_CONCENTRATION, results));
                return false;
            }
            return true;
        }

        /// <summary>
        /// Generic check dispatcher that routes different types to the appropriate check method.
        /// </summary>
        public bool Check(object obj, Hints hints = Hints.NONE)
        {
            return obj switch
            {
                AutoGainResult agr => Check(agr, hints),
                SingleMeasurement sm => Check(sm, hints),
                FirstAirMeasurementResult fam => Check(fam, hints),
                FirstSampleMeasurementResult fsm => Check(fsm, hints),
                Measurement m => Check(m, hints),
                Results r => Check(r, hints),
                _ => throw new ArgumentException($"Unsupported class {obj.GetType()}")
            };
        }

        /// <summary>
        /// Converts all entries to a JSON array.
        /// </summary>
        public JsonArray ToJson()
        {
            JsonArray array = new();
            foreach (var entry in entries)
            {
                array.Add(entry.ToJson());
            }
            return array;
        }

        /// <summary>
        /// Heuristic test for cuvette presence based on LED power and expected RFU interpolation
        /// scaled by <see cref="RfuThresholdMultiplier"/>.
        /// </summary>
        private bool HasCuvette(SingleMeasurement sm)
        {
            int ledPower = sm.Channel470.LedPower;
            if (ExpectedMaxLedPower == ExpectedMinLedPower)
                throw new InvalidOperationException("LED power interpolation division by 0");

            // Interpolated expected value
            double slope = (ExpectedMaxRfu - ExpectedMinRfu) / (ExpectedMaxLedPower - ExpectedMinLedPower);
            double expectedRfu = slope * (ledPower - ExpectedMinLedPower) + ExpectedMinRfu;

            return sm.Delta() > expectedRfu * RfuThresholdMultiplier;
        }
    }
}

File: api/python/src/hse/evifluor/verification.py
# SPDX-License-Identifier: MIT
# SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

from enum import Enum, Flag

from hse.evifluor.device import (
    AutoGainResult,
    FirstAirMeasurementResult,
    FirstSampleMeasurementResult,
)
from hse.evifluor.measurement import Measurement, Results
from hse.evifluor.singlemeasurement import SingleMeasurement


class Verification:
    """Provides verification checks for various measurement types and conditions."""
    _DEFAULT_STD_HIGH_TARGET_SIGNAL_FACTOR      = 0.8
    _DEFAULT_MIN_RFU                          = 4.5
    _min_rfu                                  = _DEFAULT_MIN_RFU
    _DEFAULT_MAX_RFU                          = 35.0
    _max_rfu                                  = _DEFAULT_MAX_RFU
    _DEFAULT_MIN_LED                          = 32
    _min_led                                  = _DEFAULT_MIN_LED
    _DEFAULT_MAX_LED                          = 222
    _max_led                                  = _DEFAULT_MAX_LED
    _DEFAULT_THRESHOLD_MULTIPLIER             = 2.0
    _threshold_multiplier                     = _DEFAULT_THRESHOLD_MULTIPLIER
    _DEFAULT_MAX_SIGNAL                       = 2499.0
    _max_signal                               = _DEFAULT_MAX_SIGNAL
    _DEFAULT_STD_HIGH_DELTA                   = 300
    _std_high_delta                           = _DEFAULT_STD_HIGH_DELTA
    _DEFAULT_THRESHOLD_NEGATIVE_CONCENTRATION = -0.1
    _threshold_negative_concentration         = _DEFAULT_THRESHOLD_NEGATIVE_CONCENTRATION

    @staticmethod
    def set_min_rfu(value):
        """Sets the minimum expected RFU value or resets to the default.

        Args:
            value: Minimum expected RFU, or ``None`` to restore the default.
        """
        if value is None:
            Verification._min_rfu = Verification._DEFAULT_MIN_RFU
        else:
            Verification._min_rfu = value

    @property
    def min_rfu(self):
        """Gets the current minimum expected RFU value."""
        return Verification._min_rfu

    @min_rfu.setter
    def min_rfu(self, value):
        """Sets the current minimum expected RFU value.

        Args:
            value: Minimum expected RFU, or ``None`` to restore the default.
        """
        Verification.set_min_rfu(value)

    @staticmethod
    def set_max_rfu(value):
        """Sets the maximum expected RFU value or resets to the default.

        Args:
            value: Maximum expected RFU, or ``None`` to restore the default.
        """
        if value is None:
            Verification._max_rfu = Verification._DEFAULT_MAX_RFU
        else:
            Verification._max_rfu = value

    @property
    def max_rfu(self):
        """Gets the current maximum expected RFU value."""
        return Verification._max_rfu

    @max_rfu.setter
    def max_rfu(self, value):
        """Sets the current maximum expected RFU value.

        Args:
            value: Maximum expected RFU, or ``None`` to restore the default.
        """
        Verification.set_max_rfu(value)

    @staticmethod
    def set_min_led(value):
        """Sets the minimum expected LED power or resets to the default.

        Args:
            value: Minimum expected LED power, or ``None`` to restore the
                default.
        """
        if value is None:
            Verification._min_led = Verification._DEFAULT_MIN_LED
        else:
            Verification._min_led = value

    @property
    def min_led(self):
        """Gets the current minimum expected LED power."""
        return Verification._min_led

    @min_led.setter
    def min_led(self, value):
        """Sets the current minimum expected LED power.

        Args:
            value: Minimum expected LED power, or ``None`` to restore the
                default.
        """
        Verification.set_min_led(value)

    @staticmethod
    def set_max_led(value):
        """Sets the maximum expected LED power or resets to the default.

        Args:
            value: Maximum expected LED power, or ``None`` to restore the
                default.
        """
        if value is None:
            Verification._max_led = Verification._DEFAULT_MAX_LED
        else:
            Verification._max_led = value

    @property
    def max_led(self):
        """Gets the current maximum expected LED power."""
        return Verification._max_led

    @max_led.setter
    def max_led(self, value):
        """Sets the current maximum expected LED power.

        Args:
            value: Maximum expected LED power, or ``None`` to restore the
                default.
        """
        Verification.set_max_led(value)

    @staticmethod
    def set_threshold_multiplier(value):
        """Sets the RFU threshold multiplier used for cuvette detection.

        Args:
            value: Threshold multiplier, or ``None`` to restore the default.
        """
        if value is None:
            Verification._threshold_multiplier = Verification._DEFAULT_THRESHOLD_MULTIPLIER
        else:
            Verification._threshold_multiplier = value

    @property
    def threshold_multiplier(self):
        """Gets the RFU threshold multiplier used when detecting cuvette presence."""
        return Verification._threshold_multiplier

    @threshold_multiplier.setter
    def threshold_multiplier(self, value):
        """Sets the RFU threshold multiplier used for cuvette detection.

        Args:
            value: Threshold multiplier, or ``None`` to restore the default.
        """
        Verification.set_threshold_multiplier(value)

    @staticmethod
    def set_max_signal(value):
        """Sets the maximum allowed signal value before saturation.

        Args:
            value: Maximum signal threshold, or ``None`` to restore the default.
        """
        if value is None:
            Verification._max_signal = Verification._DEFAULT_MAX_SIGNAL
        else:
            Verification._max_signal = value

    @property
    def max_signal(self):
        """Gets the maximum allowed signal value before saturation."""
        return Verification._max_signal

    @max_signal.setter
    def max_signal(self, value):
        """Sets the maximum allowed signal value before saturation.

        Args:
            value: Maximum signal threshold, or ``None`` to restore the default.
        """
        Verification.set_max_signal(value)

    @staticmethod
    def set_std_high_delta(value):
        """Sets the allowed deviation around the standard-high target.

        Args:
            value: Allowed deviation, or ``None`` to restore the default.
        """
        if value is None:
            Verification._std_high_delta = Verification._DEFAULT_STD_HIGH_DELTA
        else:
            Verification._std_high_delta = value

    @property
    def std_high_delta(self):
        """Gets the allowed deviation around the standard high target."""
        return Verification._std_high_delta

    @std_high_delta.setter
    def std_high_delta(self, value):
        """Sets the allowed deviation around the standard-high target.

        Args:
            value: Allowed deviation, or ``None`` to restore the default.
        """
        Verification.set_std_high_delta(value)

    @staticmethod
    def set_threshold_negative_concentration(value):
        """Sets the threshold used to flag negative concentrations.

        Args:
            value: Negative concentration threshold, or ``None`` to restore the
                default.
        """
        if value is None:
            Verification._threshold_negative_concentration = Verification._DEFAULT_THRESHOLD_NEGATIVE_CONCENTRATION
        else:
            Verification._threshold_negative_concentration = value

    @property
    def threshold_negative_concentration(self):
        """Gets the threshold used to flag negative concentrations."""
        return Verification._threshold_negative_concentration

    @threshold_negative_concentration.setter
    def threshold_negative_concentration(self, value):
        """Sets the threshold used to flag negative concentrations.

        Args:
            value: Negative concentration threshold, or ``None`` to restore the
                default.
        """
        Verification.set_threshold_negative_concentration(value)

    class ProblemId(Enum):
        """Problem identifiers tracked during verification."""
        SATURATION             = 1
        CUVETTE_MISSING        = 2
        MIN_LED_POWER          = 3
        MAX_LED_POWER          = 4
        AUTO_GAIN_RESULT       = 5
        WRONG_LEVEL            = 6
        NEGATIVE_CONCENTRATION = 7
    
    class Hints(Flag):
        """Context-sensitive hints that adjust verification checks."""
        NONE                   = 0
        MUST_HAVE_CUVETTE      = 1
        STD_HIGH               = 2

    class Entry:
        def __init__(self, problem_id, data):
            """Initializes a verification entry.

            Args:
                problem_id: Specific problem identifier.
                data: Data associated with the problem, such as a measurement or
                    result object.
            """
            self.problem_id = problem_id
            self.data       = data
    
        def __repr__(self):
            """Returns a readable representation of the entry."""
            return "{}({}) {}".format(self.problem_id.name, self.problem_id.value, self.data)
    
        def to_json(self):
            """Converts the entry to a JSON-friendly dictionary."""
            ret = { 'problem_id'  : self.problem_id.value,
                    'description' : self.problem_id.name}
            if callable(getattr(self.data, "to_json", None)):
                ret['data'] = self.data.to_json()
            else:
                ret['data'] = self.data
            return ret

        @staticmethod
        def from_json(node):
            """Restores a verification entry from JSON data.

            Args:
                node: JSON object containing serialized entry data.

            Returns:
                A reconstructed :class:`Verification.Entry`.
            """
            return Verification.Entry(Verification.ProblemId(node['problem_id']), node['data'])

    def __init__(self):
        """Initializes an empty verification result container."""
        self.entries = []
        return
        
    def __repr__(self):
        """Returns a textual representation of all collected entries."""
        return "entries:{}".format(self.entries)
        
    def success(self):
        """Indicates whether verification completed without findings."""
        return len(self.entries) == 0
        
    def failed(self):
        """Indicates whether any verification issues were recorded."""
        return len(self.entries) > 0
        
    def has_problem(self, problem_id):
        """Checks whether a specific problem has already been recorded.

        Args:
            problem_id: Problem identifier to look for.

        Returns:
            ``True`` if the problem is already present, otherwise ``False``.
        """
        for entry in self.entries:
            if entry.problem_id == problem_id:
                return True
        return False

    def add_problem_id(self, problem_id, data):
        """Adds a problem entry if it has not been recorded already.

        Args:
            problem_id: Problem identifier to add.
            data: Data associated with the detected problem.
        """
        if(not self.has_problem(problem_id)):
            self.entries.append(self.Entry(problem_id, data))
        
    def check_auto_gain_result(self, auto_gain_result, hints = None):
        """Checks the result of the auto-gain procedure for errors.

        Args:
            auto_gain_result: Auto-gain result to verify.
            hints: Optional verification hints. Currently accepted for API
                symmetry but not used.

        Returns:
            ``True`` if the result passes verification, otherwise ``False``.
        """
        ret = True
        if not (auto_gain_result.found == True):
            self.add_problem_id(self.ProblemId.AUTO_GAIN_RESULT, auto_gain_result)
            ret = False

        return ret

    @staticmethod
    def expected_value(led_power):
        """Computes the expected RFU value for a given LED power.

        Args:
            led_power: LED power level for which the expected RFU is calculated.

        Returns:
            The interpolated expected RFU value.
        """
        slope = (Verification._max_rfu - Verification._min_rfu) / (Verification._max_led - Verification._min_led)
        return Verification._min_rfu + slope * (led_power - Verification._min_led)

    @staticmethod
    def has_cuvette(sm: SingleMeasurement) -> bool:
        """Heuristic test for cuvette presence based on LED power and RFU."""
        expected = Verification.expected_value(sm.channel_470.led_power)
        return sm.delta() > expected * Verification._threshold_multiplier

    def check_single_measurement(self, single_measurement, hints = None, std_high_target_signal_factor = _DEFAULT_STD_HIGH_TARGET_SIGNAL_FACTOR):      
        """Checks a single measurement for saturation and expected behavior.

        Args:
            single_measurement: Measurement to verify.
            hints: Optional verification hints such as cuvette or standard-high
                expectations.
            std_high_target_signal_factor: Target factor applied to the maximum
                signal when checking standard-high measurements.

        Returns:
            ``True`` if the measurement passes verification, otherwise ``False``.
        """
        ret = True

        if std_high_target_signal_factor is None:
            std_high_target_signal_factor = Verification._DEFAULT_STD_HIGH_TARGET_SIGNAL_FACTOR
        
        if single_measurement.channel_470.value >= self._max_signal:
            self.add_problem_id(self.ProblemId.SATURATION, single_measurement)
            ret = False            

        if hints != None and self.Hints.MUST_HAVE_CUVETTE in hints:
            if not (Verification.has_cuvette(single_measurement) == True):
                self.add_problem_id(self.ProblemId.CUVETTE_MISSING, single_measurement)
                ret = False
                
        if hints != None and self.Hints.STD_HIGH in hints:
            std_high_target = Verification._max_signal * std_high_target_signal_factor
            if not (single_measurement.channel_470.value >= (std_high_target - Verification._std_high_delta) and single_measurement.channel_470.value <= (std_high_target + Verification._std_high_delta)):
                self.add_problem_id(self.ProblemId.WRONG_LEVEL, single_measurement)
                ret = False

        return ret
        
    def check_measurement(self, measurement, hints = None):
        """Verifies a full air/sample measurement pair.

        Args:
            measurement: Full measurement containing air and sample values.
            hints: Optional verification hints applied to the sample.

        Returns:
            ``True`` if both measurements pass verification, otherwise ``False``.
        """
        ret1 = self.check_single_measurement(measurement.air, self.Hints.MUST_HAVE_CUVETTE )
        ret2 = self.check_single_measurement(measurement.sample, (hints or self.Hints.NONE) | self.Hints.MUST_HAVE_CUVETTE)
        return ret1 and ret2

    def check_result(self, results, hints = None):
        """Validates computed results against negative concentration thresholds.

        Args:
            results: Computed concentration results to verify.
            hints: Optional verification hints. Currently unused.

        Returns:
            ``True`` if the result passes verification, otherwise ``False``.
        """
        if results.concentration < Verification._threshold_negative_concentration:
            self.add_problem_id(self.ProblemId.NEGATIVE_CONCENTRATION, results)
            ret = False
        else:
            ret = True
        return ret

    def check_first_air_measurement_result(self, fam, hints):
        """Verifies a first-air measurement result.

        Args:
            fam: First-air measurement result containing min/max measurements.
            hints: Optional verification hints. Currently unused.

        Returns:
            ``True`` if both contained measurements pass verification.
        """
        ret1 = self.check_single_measurement(fam.min_measurement, self.Hints.MUST_HAVE_CUVETTE)
        ret2 = self.check_single_measurement(fam.max_measurement, self.Hints.MUST_HAVE_CUVETTE)
        return ret1 and ret2
        
    def check_first_sample_measurement_result(self, fsm, hints, std_high_target_signal_factor):
        """Verifies a first-sample measurement result.

        Args:
            fsm: First-sample measurement result to verify.
            hints: Optional verification hints passed to the auto-gain check.
            std_high_target_signal_factor: Target factor applied during the
                standard-high level check.

        Returns:
            ``True`` if both auto-gain and sample measurement pass verification.
        """
        ret1 = self.check_auto_gain_result(fsm.auto_gain_result, hints)
        ret2 = self.check_single_measurement(fsm.measurement,  self.Hints.MUST_HAVE_CUVETTE | self.Hints.STD_HIGH, std_high_target_signal_factor = std_high_target_signal_factor)
        return ret1 and ret2
        
    def check(self, sample, hints = None, std_high_target_signal_factor = _DEFAULT_STD_HIGH_TARGET_SIGNAL_FACTOR):
        """Dispatches verification based on the provided object type.

        Args:
            sample: Object to verify. Supported types include
                :class:`SingleMeasurement`, :class:`AutoGainResult`,
                :class:`FirstAirMeasurementResult`,
                :class:`FirstSampleMeasurementResult`, :class:`Measurement`, and
                :class:`Results`.
            hints: Optional verification hints.
            std_high_target_signal_factor: Target factor applied during
                standard-high verification.

        Returns:
            ``True`` if the object passes verification, otherwise ``False``.
        """
        if isinstance(sample, SingleMeasurement):
            return self.check_single_measurement(sample, hints, std_high_target_signal_factor = std_high_target_signal_factor)
        elif isinstance(sample, AutoGainResult):
            return self.check_auto_gain_result(sample, hints)
        elif isinstance(sample, FirstAirMeasurementResult):    
            return self.check_first_air_measurement_result(sample, hints)
        elif isinstance(sample, FirstSampleMeasurementResult):    
            return self.check_first_sample_measurement_result(sample, hints, std_high_target_signal_factor)
        elif isinstance(sample, Measurement):    
            return self.check_measurement(sample, hints)
        elif isinstance(sample, Results):
            return self.check_result(sample, hints)
        else:
            raise Exception("Unsupported class {}".format(type(sample)))
        
    def to_json(self):
        """Serializes all verification entries to a JSON-compatible list."""
        ret = []
        for entry in self.entries:
            ret.append(entry.to_json())
        return ret
        
    @staticmethod
    def from_json(node):
        """Restores verification entries from JSON.

        Args:
            node: JSON array containing serialized verification entries.

        Returns:
            A reconstructed :class:`Verification` instance.
        """
        ret = Verification()
        for entry in node:
            ret.entries.append(Verification.Entry.from_json(entry))
        return ret
